duplex_http_call.go

```diff
@@ -77,17 +78,33 @@ func newDuplexHTTPCall(
 		Proto:      "HTTP/1.1",
 		ProtoMajor: 1,
 		ProtoMinor: 1,
-		Body:       pipeReader,
 		Host:       url.Host,
+		Body:       pipeReader,
+		GetBody: func() (io.ReadCloser, error) {
+			// GetBody is called by the http client on request retries.
+			// We need to return a reader that will read from the buffer
+			// and then the pipe reader.
+			buf := writeBuffer.getBytes()
```

Comment:
I don't think this is safe or correct. We are assuming that everything written to the pipe has also been written to the network of the previous attempt. So we're always replaying everything written.

However, what we _really_ need to be monitoring is everything **read** from the pipe. Those are the bytes that are actually _absent_ from the pipe reader that need to be "re-played". There's no guarantee that bytes are copied from the pipe reader to the network in chunks that are the same size as the bytes written to the pipe so far.

A write to a pipe writer _blocks_ until all of the bytes have been read. So what could happen is that a write could be in-progress and only partially written to the network. So the buffer won't have the full copy of that in-progress message. When we then combine the buffer (any previously written messages) and the pipe reader (which has only a _portion_ of the in-progress message), we end up losing bytes and writing a corrupt request body.

Reply:
I think this is fixed. Thanks for catching! Will write some tests for it.

Reply:
I've fixed but it's racy on claiming and releasing the buffer. Needs more work.

Reply:
New `messagePipe` works on complete writes to avoid buffering until needed.